RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1481 @@
+            auto const tmpmcol = valueM_col - value_col;
+
+            // diagonal
+            auto tmp_value = 0.5 * (tmppcol * tmppcol + tmpmcol * tmpmcol);
+            covarianceMatrix(col, col) += ampl2 * tmp_value;
+
+            // FIXME: understand if this actually gets unrolled
+            CMS_UNROLL_LOOP
+            for (int row = col + 1; row < NSAMPLES; row++) {
+              float const valueP_row = pmpcol[row];  //pulseMatrixP(j, ipulseReal);
+              float const value_row = pmcol[row];    //pulseMatrix(j, ipulseReal);
+              float const valueM_row = pmmcol[row];  //pulseMatrixM(j, ipulseReal);
+
+              float tmpprow = valueP_row - value_row;
+              float tmpmrow = valueM_row - value_row;
+
+              auto const covValue = 0.5 * (tmppcol * tmpprow + tmpmcol * tmpmrow);
+
+              covarianceMatrix(row, col) += ampl2 * covValue;
+            }
+          }
+        }
+      }
+
+      template <int NSAMPLES, int NPULSES>
+      class kernel_minimize {
+      public:
+        template <typename TAcc, typename = std::enable_if_t<alpaka::isAccelerator<TAcc>>>
+        ALPAKA_FN_ACC void operator()(TAcc const& acc,
+                                      OProductType::View outputGPU,
+                                      float const* amplitudes,
+                                      float* pulseMatrices,
+                                      float* pulseMatricesM,
+                                      float* pulseMatricesP,
+                                      HcalMahiPulseOffsetsPortableDevice::ConstView pulseOffsetsView,
+                                      float* noiseTerms,
+                                      float* electronicNoiseTerms,
+                                      float* soiSamples,
+                                      HcalMahiConditionsPortableDevice::ConstView mahi,
+                                      bool const useEffectivePedestals,
+                                      IProductTypef01::ConstView f01HEDigis,
+                                      IProductTypef5::ConstView f5HBDigis,
+                                      IProductTypef3::ConstView f3HBDigis) const {
+          // can be relaxed if needed - minor updates are needed in that case!
+          static_assert(NPULSES == NSAMPLES);
+
+          auto const nchannels = f01HEDigis.size() + f5HBDigis.size() + f3HBDigis.size();
+
+          auto const nchannelsf015 = f01HEDigis.size() + f5HBDigis.size();
+
+          auto const threadsPerBlock(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[0u]);
+
+          //Loop over all groups of channels
+          for (auto group : uniform_groups_x(acc, nchannels)) {
+            //Loop over each channel
+            for (auto channel : uniform_group_elements_x(acc, group, nchannels)) {
+              auto const gch = channel.global;
+              auto const lch = channel.local;
+
+              if (gch >= nchannels)
+                return;
```

Comment:
Shouldn't `uniform_group_elements_x()` guarantee `gch < nchannels`?

Reply:
removed https://github.com/cms-sw/cmssw/commit/98367cbd606553c45f30a100a438bedbe880ae55

Reply:
Mhm, according to GitHub this is till there ?